### DISCUSSION TITLE
How to tell that you're working with a CALM file

### DISCUSSION THREAD

**Discussion by aidanm3341** [finos/architecture-as-code#482] [Ideas]
For the purposes of the VSCode Extension and other tooling, we need a way to know that the a given file is intended to be using CALM. We could look for a reference to the schema, but this doesn't always work given that a CALM instance isn't required to be based on a pattern, and therefore doesn't need to reference any external file at all that we could look for containing 'calm'. 

I would propose that we have CALM files include it in their file name, e.g. `api-gateway.calm.json`. It would also be useful to specify a pattern this way, possibly like this - `api-gateway.calm-pattern.json`. This gives us a clear convention that we can check for in our tooling to know what we're working with, and in general makes it more obvious to users what kinds of files they have.

What do people think of this approach?

**Comment by rocketstack-matt** [comment]
Given CALM isn't a single schema, but a base meta schema and then a bunch of additional schemas including custom schemas that a user can provide themselves (patterns, control requirements) what does this actually solve? Aren't you still going to have to look inside the file to figure out what the schema is?

**Reply by aidanm3341** [reply]
take this example https://github.com/finos/architecture-as-code/blob/main/calm/samples/api-gateway-instantiation.json

how would you know looking at this that you're working with CALM?

EDIT: unless of course this is an invalid instance... I've been assuming that it's valid. Is the `$schema` _required_ in every instance? If so, are you suggesting that we just check for the existence of the CALM schema being referenced? e.g. like this example https://github.com/finos/architecture-as-code/blob/main/calm/samples/traderx/traderx.json

**Reply by rocketstack-matt** [reply]
That's invalid, there should be a schema reference . . . however, per my point above the schema won't necessarily be the calm schema, if you're working on an instantiation of a pattern for example then the schema ref would be to the pattern, if you're writing a control-configuration the the schema would be the control-requirement.

**Reply by aidanm3341** [reply]
right. So we should always be able to follow those schema references until we reach the core schema, and if we are unable to do that then it must not be CALM

**Reply by Thels** [reply]
I'll raise an issue to review the pre-existing files checked into the repository to validate that they are suiting their purpose.

In terms of patterns/instantiations -  I just want to confirm what the expected parsing behaviour is. I know this will need to change in the future for validating other items like controls/flows etc.

If the current file has a `$schema` contains `/pattern/` then it's very likely that you're in an instantiation - with the pattern being that given value.

Otherwise if the `$id` contains `/pattern/` you're likely parsing a pattern itself.

Then the version of CALM to validate it against ( instead of having this hard-coded anywhere in the code like it currently is ) we can simply go through what data we have available ( pattern / instantiation ) to find a reference to calm.json.

**Reply by Thels** [reply]
https://github.com/finos/architecture-as-code/issues/484

Raised

**Reply by rocketstack-matt** [reply]
> I'll raise an issue to review the pre-existing files checked into the repository to validate that they are suiting their purpose.
> 
> In terms of patterns/instantiations - I just want to confirm what the expected parsing behaviour is. I know this will need to change in the future for validating other items like controls/flows etc.
> 
> If the current file has a `$schema` contains `/pattern/` then it's very likely that you're in an instantiation - with the pattern being that given value.
> 
> Otherwise if the `$id` contains `/pattern/` you're likely parsing a pattern itself.
> 
> Then the version of CALM to validate it against ( instead of having this hard-coded anywhere in the code like it currently is ) we can simply go through what data we have available ( pattern / instantiation ) to find a reference to calm.json.

This may be hard to enforce, there is nothing preventing you using a URI that does not have the word pattern in; it's a nice convention but nothing more.

**Reply by Thels** [reply]
Completely agree, this is just for the initial implementation of the extension.  We need to have an easier way to derive this context from the file we're processing.

Could we potentially introduce another element into the `core` schema?  I may be massively over-simplifying the problem space though.

`"calm-type": ["pattern", "pattern-instantiation", ...others ]`

**Reply by willosborne** [reply]
This is the approach that Angular uses - `.service.ts`, `.component.ts`, etc - and they go further with `.spec.ts` for tests.
It's really clear, separates things out in the editor and is incredibly easy to work with from a tooling perspective without having to expand the spec.

I guess for CALM this would look like
- `.calm.json` for CALM architectures/instantiations - this is the base document 
- `.pattern.json` for patterns
- `.requirement.json` or similar for others? `.calm.requirement.json` feels a bit excessive

**Reply by Thels** [reply]
I really like this idea, it would give a clear way to know what you expect your current file to contain without parsing the contents.

**Comment by aidanm3341** [comment]
Despite establishing that we probably _could_ determine if we're working with CALM based on the file contents, I wonder if it's still worth introducing this just as a naming convention anyway. 

Taking again the [trader x example](https://github.com/finos/architecture-as-code/blob/main/calm/samples/traderx/traderx.json), only someone with experience with CALM could look at this and know that it's an instantiation and not a pattern. If we introduce a naming convention, at least someone who is only familiar with the concepts could look at this and understand what they have. I'm thinking more just from a user experience perspective now

**Reply by Thels** [reply]
But isn't that example incorrect like the one you referenced earlier in this discussion? Shouldn't the `$schema` refer to the **pattern**   it was created from?

**Reply by aidanm3341** [reply]
in this case, it wasn't created from a pattern. The traderx example is just a standalone architecture

**Reply by Thels** [reply]
Ah okay fair point, from looking at the validation code in the CLI, it looks like assumptions were made that all instantiations refer to a pattern.  Potentially we need to introduce the concept of validating stand-alone?

**Reply by aidanm3341** [reply]
yes agreed, this is a limitation of the current validator that we should address

**Reply by Thels** [reply]
Raised an issue to suggest addressing this.

**Reply by LeighFinegold** [reply]
What was the  issue number?
